crates/uv-resolver/src/pubgrub/dependencies.rs

```diff
@@ -138,12 +141,25 @@ fn to_pubgrub(
 
         // The requirement has a specifier (e.g., `flask>=1.0`).
         Some(VersionOrUrl::VersionSpecifier(specifiers)) => {
-            let version = specifiers
-                .iter()
-                .map(PubGrubSpecifier::try_from)
-                .fold_ok(Range::full(), |range, specifier| {
-                    range.intersection(&specifier.into())
-                })?;
+            // If the specifier is an exact version, and the user requested a local version that's
+            // more precise than the specifier, use the local version instead.
+            let version = if let Some(expected) = locals.get(&requirement.name) {
```

Comment:
`locals.get` should be free (the hash map short-circuits if it's empty) unless local dependencies were declared.